readthedocs/projects/tasks/builds.py

```diff
@@ -263,25 +265,12 @@ def _check_concurrency_limit(self):
             max_concurrent_builds = settings.RTD_MAX_CONCURRENT_BUILDS
 
         if concurrency_limit_reached:
-            # TODO: this could be handled in `on_retry` probably
-            log.warning(
-                'Delaying tasks due to concurrency limit.',
-                project_slug=self.data.project.slug,
-                version_slug=self.data.version.slug,
-            )
-
-            # This is done automatically on the environment context, but
-            # we are executing this code before creating one
-            api_v2.build(self.data.build['id']).patch({
-                'error': BuildMaxConcurrencyError.message.format(
+            # By raising this exception and using ``autoretry_for``, Celery
+            # will handle this automatically calling ``on_retry``
+            raise BuildMaxConcurrencyError(
+                BuildMaxConcurrencyError.message.format(
                     limit=max_concurrent_builds,
-                ),
-                'builder': socket.gethostname(),
-            })
-            self.retry(
-                exc=BuildMaxConcurrencyError,
-                # We want to retry this build more times
-                max_retries=25,
+                )
```

Comment:
Ooooh, I see what you were talking about now with `autoretry_for`. I wasn't familiar with this attribute when we were last talking about this. I like this syntax, it makes more sense :+1: